Cleanup the README.rst for the lifecycle demo. (#508)
ros2/demos: lifecycle/README.rst

Introduction
------------

ROS 2 introduces the concept of managed nodes, also called ``LifecycleNode``\ s.
In the following tutorial, we explain the purpose of these nodes, what makes them different from regular nodes and how they comply to a lifecycle management.

> Managed nodes are scoped within a state machine of a finite amount of states.

Reviewer: I don't think this sentence is very clear. Managed nodes implement a state machine. I don't know what "scoped within a state machine" is referring to and I think that a finite number of states goes without saying.

Author: I refactored it in 3556153

These states can be changed by invoking a transition id which indicates the succeeding consecutive state.
The state machine is implemented as described at the `ROS 2 design page <http://design.ros2.org/articles/node_lifecycle.html>`__.

Our implementation differentiates between ``Primary States`` and ``Transition States``.

> Primary States are supposed to be steady states in which any node can do the respected task.

Reviewer: I think this terminology is awkward. A state machine has states and transitions. I think it's potentially confusing to define a "transition state."

Author: While I agree with you in an abstract sense, we do indeed define "transition states" as part of the state machine: https://github.com/ros2/rcl_interfaces/blob/master/lifecycle_msgs/msg/State.msg .  So this sentence matches the code, even though the terminology is a bit weird.

Reviewer: Ok. Good that the docs are in sync.

#. cleanup: We destroy the publisher and timer

> The principle is implemented in this demo as the typical talker/listener demo.

Reviewer: What principle? Instead of mentioning a principle that wasn't introduced could simply say that the demo implements a lifecycle node.

> However, imagine a real scenario with attached hardware which may have a rather long booting phase, i.e. a laser or camera.

Reviewer: "However" would indicate some kind of contrast, but there isn't one. It's superfluous here and it could just say, "Imagine a real..."

Author: I didn't make any changes here, as with the previous cleanup I think this is fairly clear now.

One could imagine bringing up the device driver in the configuring state, start and stop only the publishing of the device's data in active/deactive state, and only in the cleanup/shutdown state actually shutdown the device.

The ``lifecycle_listener`` is a simple listener which shows the characteristics of the lifecycle talker.

> The ``lifecycle_talker`` enables message publishing only in the active state and thus the listener only receives messages when the talker is in an active state.

Reviewer: An aside: unfortunately, subscribers are not lifecycle enabled. This creates a bit of an issue because a lifecycle node that is not yet in the Active state can receive messages and so has to implement its own mechanism to avoid receiving messages when not in the Active state. This can happen when lifecycle nodes are mixed with other nodes (publishers) that are not lifecycle enabled.

Author: Agreed.  That problem is tracked somewhere else (I can't find the ticket at the moment), but I'm not sure we should update anything here in the documentation about that.  We'll obviously have to update this once we implement a subscriber state machine.

ros2 lifecycle command line interface
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The ``lifecycle_service_client`` application is a fixed order script for demo purposes only.

Reviewer 2: @clalancette I understand what it means by _fixed order_, but perhaps it can be phrased better?

> The ``lifecycle_service_client`` application is a fixed order script for demo purposes only.

Author: ```suggestion
The ``lifecycle_service_client`` application executes a fixed set of operations for demo purposes only.
```
